cmd/tool/tool.go

```diff
@@ -24,40 +30,115 @@ func (t CodacyTrivy) Run(tool codacy.Tool, sourceDir string) ([]codacy.Issue, er
+		case ruleIDVulnerability:
+			scanners = append(scanners, types.VulnerabilityScanner)
+		}
+	}
+
+	if len(scanners) == 0 {
+		return nil, ToolError{msg: "Failed to configure Codacy Trivy: no pattern matches existing rules"}
 	}
 
-	scanner := secret.NewScanner(&secret.Config{})
+	// The `quiet` field in global options is not used by the runner.
+	// This is the only way to suppress Trivy logs.
+	log.InitLogger(false, true)
 
-	results := []codacy.Issue{}
+	return &flag.Options{
+		GlobalOptions: flag.GlobalOptions{
+			// CacheDir needs to be explicitly set and match the directory in the Dockerfile.
+			// The cache dir will contain the pre-downloaded vulnerability DBs.
+			CacheDir: cacheDir,
+		},
+		DBOptions: flag.DBOptions{
+			// Do not try to update vulnerability DBs.
+			SkipDBUpdate:     true,
+			SkipJavaDBUpdate: true,
+		},
+		ReportOptions: flag.ReportOptions{
+			// Listing all packages will allow to obtain the line number of a vulnerability.
+			ListAllPkgs: true,
+		},
+		ScanOptions: flag.ScanOptions{
+			// Do not try to connect to the internet to download vulnerability DBs, for example.
+			OfflineScan: true,
+			Scanners:    scanners,
+			// Instead of scanning files individually, scan the whole source directory since it's faster.
+			Target: sourceDir,
+		},
+		VulnerabilityOptions: flag.VulnerabilityOptions{
+			// Only scan libraries not OS packages.
+			VulnType: []types.VulnType{types.VulnTypeLibrary},
+		},
+	}, nil
+}
+
+func run(ctx context.Context, config flag.Options) ([]codacy.Issue, error) {
+	runner, err := artifact.NewRunner(ctx, config)
+	if err != nil {
+		return nil, ToolError{msg: "Failed to initialize Codacy Trivy", w: err}
+	}
+	defer runner.Close(ctx)
+
+	results, err := runner.ScanFilesystem(ctx, config)
+	if err != nil {
+		return nil, ToolError{msg: "Failed to run Codacy Trivy", w: err}
+	}
+
+	issues := []codacy.Issue{}
+	for _, result := range results.Results {
+		// Make a package map for faster lookup
+		packagesWithLineNumberById := map[string]ftypes.Package{}
+		for _, pkg := range result.Packages {
+			// Only add packages that have a line number
+			if len(pkg.Locations) > 0 {
+				packagesWithLineNumberById[pkg.ID] = pkg
+			}
+		}
 
-	for _, f := range files {
-		content, err := os.ReadFile(path.Join(sourceDir, f))
-		if err != nil {
-			return nil, ToolError{msg: fmt.Sprintf("Failed to read source file %s", f), w: err}
+		// Vulnerability scanning results
+		for _, vuln := range result.Vulnerabilities {
+			// Only create issues that have an associated line number
+			pkg, pkgHasLineNumber := packagesWithLineNumberById[vuln.PkgID]
+			if pkgHasLineNumber {
+				issues = append(
+					issues,
+					codacy.Issue{
+						File:      result.Target,
+						Line:      pkg.Locations[0].StartLine, // Safe to access index 0 due to map construction above.
+						Message:   fmt.Sprintf("Insecure dependency %s (%s: %s) (update to %s)", vuln.PkgID, vuln.VulnerabilityID, vuln.Title, vuln.FixedVersion),
```

Comment:
we also had aligned on the format: Insecure dependency {pkg_name}**@{version}** ({vulnerability_id}: {vulnerability title}) (update to {fix_versions})

is it possible to include the impacted version? (it might not be explicit in the package manager)

Reply:
nevermind, I noticed the `vuln.PkgID` already includes the version 👍